### ISSUE TITLE
Push official containers to dockerhub

### ISSUE THREAD

**Issue by Trolldemorted** [arkime/arkime#1363]
At the moment there are no official moloch images on dockerhub, which has created the undesirable situation where a plethora of unofficial moloch images have been uploaded to dockerhub, including but not limited to:
- danielguerra/docker-moloch
- piesecurity/docker-moloch
- tfplenum/moloch

Since none of these users can be trusted to deliver patches in an acceptable timespan (or at all), and we definitely can not put personally identifiable information into a container built by some random person on the internet, it would be great if you'd publish your releases to dockerhub.

Do you have any CI/CD to build releases? Could it automatically also build and push docker images?

**Comment by awick** [comment]
We have no official docker release.  Always looking for someone to contribute and work with the community on building one. There are several folks that use docker if you want to join slack and work on creating one.

**Comment by Trolldemorted** [comment]
Is there any CI/CD infrastructure in place that we can convince to build and push?

Are the owners of this project willing to solely take care of the administrative tasks (creating an organization on dockerhub, ensuring that access to the organization is not given to third parties and nobody expect them can push new imges, reaching out if builds fail for whatever reason)? We might be able to help you with setting up the build pipeline, but access to your images should not be given to anyone outside your organization (including me).

**Comment by awick** [comment]
yes, there is CICD using screwdriver, see the screwdriver.yml

Building/uploading is the easy part.  we already use docker for the build process itself.  The hard part is getting everyone to agree on how it should be done, and since we don't use docker in production currently I don't feel like the person to do it.

**Comment by Trolldemorted** [comment]
Yeah, the worst thing you can expect from developers is to agree on something :)

After a bit of reading and playing around, I noticed two problems:
- moloch-viewer does not automatically set up its database. One might be tempted to write a docker-entrypoint.sh which contains something like this:
    ```
    if [ ! -f "/molochsetup" ]; then
        /data/moloch/db/db.pl http://elasticsearch:9200 initnoprompt
        /data/moloch/bin/moloch_add_user.sh "$MOLOCH_USER" "$MOLOCH_USER" "$MOLOCH_PASSWORD" --admin
        touch /molochsetup
    else
        /data/moloch/db/db.pl http://elasticsearch:9200 upgradenoprompt
    fi
    ```
    but this nukes the database if the moloch-viewer container is rebuilt from the image. Is there any chance that either
    - moloch-viewer takes care of initializing and upgrading the database internally and automatically, or
    - `db.pl upgradenoprompt` also does what initnoprompt does, but only if necessary?

    Most ORMs support the "your database is over there, you know all migrations, now ensure that all up to the latest are applied" approach. Is there a specific reason users have to trigger migrations manually?
- The documentation about remote sniffers is incomplete. The documentation states that ES is filled with the metadata, implying that the viewer pulls a connection's content from the corresponding sniffer, but it is not going into sufficient detail that would tell me how. There is a discussion in #55 where a user asks how to organize that, but it is lacking a definitive answer and the entire issue could be severely outdated as it is from 2013. From my point of view, the ideal setup looks like this:
    - elasticsearch and moloch-viewer are running in a container
    - moloch-captures are running on routers/bastion hosts/firewalls/wherever

    I believe (correct me if I am wrong) this implies that
    - every moloch-capture needs to be able to connect to ES
    - every moloch-capture has to declare where moloch-viewer can reach the contents
    - moloch-viewer must be able to reach the locations the captures have declared

    How does this work? This is essentially the same problem as #55 - there is no obvious way to have capture, viewer and storage running on separate systems.

Could you help us out, or tell us how you see things?

**Comment by awick** [comment]
You really should join the slack workspace, especially the #docker channel.  

Do you already run moloch?  If not I would suggest you start there.


> After a bit of reading and playing around, I noticed two problems:

problems is a strong word


> * moloch-viewer does not automatically set up its database. One might be tempted to write a docker-entrypoint.sh which contains something like this:

For this phase 1 of getting docker right it shouldn't do anything with the db automatically.


> * The documentation about remote sniffers is incomplete. The documentation states that ES is filled with the metadata, implying that the viewer pulls a connection's content from the corresponding sniffer, but it is not going into sufficient detail that would tell me how. There is a discussion in #55 where a user asks how to organize that, but it is lacking a definitive answer and the entire issue could be severely outdated as it is from 2013. From my point of view, the ideal setup looks like this:

Not sure what you are talking about here.

There are definitely 2 different designs, those runnings for a cloud environment and those not.


>   
>   * elasticsearch and moloch-viewer are running in a container

they should NOT be the same container.  Also don't want to get in the business of providing any kind of elasticsearch container etc.

>   * moloch-captures are running on routers/bastion hosts/firewalls/wherever

capture doesn't run on routers/firewalls usually

>   
>   I believe (correct me if I am wrong) this implies that
>   
>   * every moloch-capture needs to be able to connect to ES
yes

>   * every moloch-capture has to declare where moloch-viewer can reach the contents

well capture and viewer are designed to be on the same machine and access the same files, except on AWS

>   How does this work? This is essentially the same problem as #55 - there is no obvious way to have capture, viewer and storage running on separate systems.

correct, that is how it works

> 
> Could you help us out, or tell us how you see things?

who is us?

**Comment by Trolldemorted** [comment]
>Do you already run moloch? If not I would suggest you start there.

Yeah, we are giving trainees access to systems whose traffic is gathered, as they sometimes need to analyze the traffic quickly and efficiently.

>For this phase 1 of getting docker right it shouldn't do anything with the db automatically.

How else should the database be created? That is what software usually does when it starts, from apps on smartphones to serverless cloud scenarios like azure functions. We cannot take care of that manually everytime a new setup is created, and the unofficial moloch images are [struggling with the same problem](https://github.com/danielguerra69/docker-moloch/blob/master/scripts/startmoloch.sh#L24):
```sh
# intialize moloch
echo INIT | /data/moloch/db/db.pl http://elasticsearch:9200 init
/data/moloch/bin/moloch_add_user.sh admin "Admin User" THEPASSWORD --admin
/data/moloch/bin/moloch_update_geo.sh
```
This one just nukes the db everytime the container starts, which I'd like to avoid. Again, is there a specific reason upgradenoprompt does not apply the first migration?


>they should NOT be the same container. Also don't want to get in the business of providing any kind of elasticsearch container etc.

Sorry, that phrasing was clumsy: I meant "are running in a separate container", so moloch-viewer and ES should have their own containers.

>capture doesn't run on routers/firewalls 

It runs on our gateways, and I kind of like the idea (capture the traffic where it arrives, don't interfere with normal ops) - but the problem is that we had to put the viewer on the gateways too, or push the entire traffic to the host running moloch-viewer.

>well capture and viewer are designed to be on the same machine and access the same files, except on AWS

This is unfortunate, so I guess we will continue to employ network shares and whatnot if need be :|

>who is us?
You or whoever can shed some light on this :)

**Comment by awick** [comment]
> How else should the database be created? That is what software usually does when it starts, from apps on smartphones to serverless cloud scenarios like azure functions. We cannot take care of that manually everytime a new setup is created, and the unofficial moloch images are [struggling with the same problem](https://github.com/danielguerra69/docker-moloch/blob/master/scripts/startmoloch.sh#L24):

Baby steps.

Not all software does that, I've used plenty where you have to run a special script to set things up so nothing is accidentally done.  Plus many apps that do stuff automagically end up having security bugs around default accounts and other things.

> Again, is there a specific reason upgradenoprompt does not apply the first migration?

I don't understand the question.

> > well capture and viewer are designed to be on the same machine and access the same files, except on AWS
> 
> This is unfortunate, so I guess we will continue to employ network shares and whatnot if need be :|

I'm confused what you expect here.  Are you asking for a NAS/NFS solution with a central viewer?  So we would have 3 methods?  If so that would be new ask and probably should be moved outside this docker request.  So end up with
* Local (capture/viewer)
* AWS (capture/central viewer)
* NAS/NFS (capture/central viewer)

**Comment by Jc2k** [comment]
I am currently migrating an AWS style (but using a non-AWS S3) setup so joined #docker to talk about this ticket. I thought it would be good to try and summarise the discussion so far here.

@awick has said the actual act of building and shipping the images is not the problem. Theres actually more of an issue about building community consensus for best practice. How to inject config, what OS to use, etc.

It looks like screwdriver can run `docker build` and `docker push` (with some caveats - as its the usual "docker in docker" problem i think you have to use the vm executor which can be slow). I didn't find an example of this specifically, but I was able to find an example of having screwdriver trigger docker hub to do the build. If all else fails, I imagine kaniko would work in a non-vm screwdriver environnment too.

I think we all agree that the container image should use the existing build shell script. It's what the apt and rpm packages already use, and means that version bumps to things like glib and yara get applied to the container directly. It massively reduces the overhead of shipping the images. This does limit the base OS's we could support for Docker - i don't think that script currently supports alpine, for example. Unless we add support to that script, of course.

I think we should use ubuntu 18.04 as the base image. Whilst we have others we like (myself included) 18.04 seems like a good compromise. We already support it for apt packages and in the existing build script. It's size is much reduced since the previous LTS images. It's a common distribution that people are familiar with. It's not like a language runtime image like `python` for example, so I think the use cases where this is deeply problematic are probably quite small. At this **early** stage no one has disagreed.

I'd prefer to not have multiple base OS's, but the build script approach means that if people did the work to support/test/whatever this wouldn't be too malodorous.

One contentious issue is getting configuration in. There is an argument that this can be left to the user and they simply use bind mounts to inject a config. But it would be nice if there was some tooling for the "easy" case.

There is already a `Configure` script so one could:

```
docker-compose run capture Configure
```

And follow the prompts to configure a standalone instance. As long as the config file was on a bind mount or volume this would be fine.

A more refined version of this is that we could either modify Configure or add a non-interactive version. Such a tool could take settings from the environment/command line. It could then be run before moloch starts to ensure the config is good. I think this might be a second babystep myself.

We talked about how k8s users can then ignore all this helper tooling and just put a config file in a `ConfigMap` or `Secret` then spin up a bunch of capture agents with a `Daemonset`. For those users having a container with the binaries is probably enough actually.

Eventually there may be 3 pcap options - local, nfs, s3. This shouldn't require a different container in terms of binaries though. The suggested config script could handle all of these within one container. And again, advance users can opt out and just provide a full config via a mount.

There was a suggestion that some users want to slice their persistent storage in different ways. Should there be one volume or multiple?

I think we can make a default structure should be able to support both. For example the completely ficticious:

```
/data/moloch/var/
    pcaps/
    folderb/
    config/
```

The container doesn't care whether you bind mount the whole of `/data/moloch/var` or make a seperate bind mount for each subfolder. So I think we should be good on that front too?

I think that's everything we've discussed so far. I'm with @awick on the db init. It would be nice if it was automatic, but babysteps. Lets not rush into the deep end and accidentally drop someones ES on the floor.

Would love to have some more discussions on #docker and see if we can reach consensus.

**Comment by rodgermoore** [comment]
Hi @Trolldemorted. I noticed your comment:

> moloch-viewer does not automatically set up its database. One might be tempted to write a docker-entrypoint.sh which contains something like this:

```
if [ ! -f "/molochsetup" ]; then
    /data/moloch/db/db.pl http://elasticsearch:9200 initnoprompt
    /data/moloch/bin/moloch_add_user.sh "$MOLOCH_USER" "$MOLOCH_USER" "$MOLOCH_PASSWORD" --admin
    touch /molochsetup
else
    /data/moloch/db/db.pl http://elasticsearch:9200 upgradenoprompt
fi
```

> but this nukes the database if the moloch-viewer container is rebuilt from the image. Is there any chance that either

I ran into this problem too and I fixed this problem by checking if the default Moloch ES indexes exist during container initialisation:

```
#Configure Moloch
echo "Step 1, configure Moloch to run."
if [ ! -f /data/configured ]; then
  touch /data/configured
  chmod +x /data/moloch/bin/Configure
  /data/moloch/bin/Configure
fi

echo "Step 2"
echo "Is Elasticsearch initialised for moloch? If not, lets do it!"
if [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/dstats_v4") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/users_v7") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/fields_v3") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/queries_v3") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/stats_v4") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/sequence_v3") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/hunts_v2") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/files_v6") == *"200"* ]] && \
   [[ ! $(curl -u elastic:YourPass --cacert /ssl/ca.crt -sl --head "https://localhost:9200/lookups_v1") == *"200"* ]]; then
   echo INIT | /data/moloch/db/db.pl https://elastic:YourPass@localhost:9200 init
   curl -u elastic:YourPass --cacert /ssl/ca.crt -sS -XPUT 'https://localhost:9200/_all/_settings?preserve_existing=true' -H 'Content-Type: application/json' -d '{"index.number_of_replicas" : "0"}'
   /data/moloch/bin/moloch_add_user.sh admin "Admin User" YourPass --admin
else echo "You're good to go!"
fi
echo "Step 3"
echo "Launch viewer..."
/bin/sh -c 'export NODE_EXTRA_CA_CERTS=/ssl/ca.crt; cd /data/moloch/viewer; /data/moloch/bin/node viewer.js -c /data/moloch/etc/config.ini | tee -a /data/moloch/logs/viewer.log 2>&1'
``` 
Hope this helps?
Currently working on implementing "Docker Secrets" using Swarm and docker-compose to move away from storing sensitive info in environmental variables.  I was wondering if and how this can be used for the Moloch config.ini file as well...? I'm on slack, find me in the Moloch #docker channel: rodgermoore.

Cheers

**Comment by Jc2k** [comment]
Continuing the thread of trying to figure out how Docker and moloch work together, I've tried to write up some common approaches to configuration with Docker and what they might mean for moloch [here](https://gist.github.com/Jc2k/eaf16b6fc9158da1bcbaee0a473725a8). 

I have successfully used the confd approach with moloch and Docker (no swarm) before and am also currently using the bind-mount approach with Kubernetes too (with `ConfigMap`/`Secret`). I have used all of these in one form or another with other stacks.

I started writing this up before your comment @rodgermoore. I've used Swarm a bit with secrets but not for any great length of time and nearly a year ago. On Kubernetes i've just put my entire ini in a kubernetes secret. But if you want to break up your config then the other option is the `includes` setting in `config.ini` - you could break up your config.ini between a config and secret that way. Or even make a custom ini with some of your config, then have that load your environment specific stuff from a ini fragment in a config and an ini fragment in a secret. Does something like that make sense for your environment?

**Comment by rodgermoore** [comment]
@Jc2k I fixed it by "generating" the config.ini file on-the-fly during start of the container (using the entrypoint script).

**Comment by Trolldemorted** [comment]
>I don't understand the question.

Sorry for the long delay @awick. If I understand the documentation correctly, upgradenoprompt applies all pending migrations to the database if and only if the database is not fresh (this implies that it will happily apply all migrations except the very first). If it would always apply all pending migrations, we could use it to initialize the database fire-and-forget style. Is there a technical reason that upgradenoprompt cannot work with an empty database?

Thanks for the suggestion @rodgermoore! I tollaly agree moving to secrets over env vars is the right way, but as long as https://github.com/containers/libpod/issues/3264 is not fixed depending on secrets essentially means you can't run your image on podman, so we won't do that yet.

I guess the cheapest solution would be starting a moloch-view and elasticsearch container, manually setting things up, `docker commit`ing it into images, and pushing those holy bespoken images to the registry of your choice...

**Comment by awick** [comment]
It is my view that db.pl should be run with intention and on purpose.  It should NOT be run every time a docker container starts.  Most folks either have multiple ES nodes or multiple capture nodes, and if there is a mass restart you don't want them all doing it.

You are correct there is no technical reason it couldn't do an init if it didn't exist other then I don't think it should.  I guess someone could also submit a initorupgradenoprompt action PR.  Although at the same time you could easily write a shell script that just detects the state and calls db.pl with the action of your desire.

**Comment by Jc2k** [comment]
Have been running an image made from the easybuttonbuild that is compatible with what i say in this ticket and [here](https://gist.github.com/Jc2k/eaf16b6fc9158da1bcbaee0a473725a8) for a while now. Pretty happy with the approach on my end. No one has shot down what i've been pushing for here or on slack, though some people have asked for more automation out of the box. I think it's fine to start simple and address the other requests as seperate tasks, though.

What do we need to do to move forward?

I could do a documentation PR to start with? It could include building the image without any moloch CD changes to start with, then we could iterate from there, and then add the image to screwdriver. Or we could update screwdriver first and then i'll document how i use that image and also some otherways you can use it.

Either way we could choose to not publically reference the container stuff until its ready - so image could be available but not referenced on releases page. And docs could be kept out sitemap and have a warning that it's not ready for production until we've tested them out a bit.

I've been milling around on slack for a while now - it's safe to say i'm not going anywhere and i'd be planning to try and help out in #docker there, if that helps.

**Comment by awick** [comment]
Completed this starting with 5.5.0, feel free to provide suggests/fixes.

https://arkime.com/docker
